### ISSUE TITLE
Add a lazy way to set Copy#destinationDir

### ISSUE THREAD

**Issue by liutikas** [gradle/gradle#25824]
### Expected Behavior

Add a lazy way to set Copy#destinationDir

### Current Behavior (optional)

Currently, the setting Copy task destination means calling setDestinationDir that takes a file. If you have a provider, you have no way to use it for set up.

### Context

As Project.buildDir is deprecated, the suggested path is to use Project.layout.buildDirectory.file("foo"), but you are unable to use that on a Copy task in a lazy way.

**Comment by liutikas** [comment]
Ideally, it just mirrors `AbstractArchiveTask#setDestinationDirectory`

https://cs.android.com/android-studio/gradle/+/master:subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java;l=142

**Comment by asodja** [comment]
I agree, while this is not the case you can use [Copy.into](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:into(java.lang.Object)). It should work with lazy types.

Note that `Copy.destinationDir` actually [calls into()](https://github.com/gradle/gradle/blob/9f50bb73fad24f9f3616285f12037e85e7650219/subprojects/core/src/main/java/org/gradle/api/tasks/Copy.java#L106-L108).

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

**Comment by liutikas** [comment]
Any updates on this FR?

**Comment by liutikas** [comment]
`Sync` task could use this as well.

**Comment by liutikas** [comment]
> I agree. Until this is not resolved you can use [Copy.into()](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:into(java.lang.Object)). It should work with lazy types.

It does not work for lazy set up of using Copy/Sync destinationDir lazily.

E.g. if you have a task that takes `Directory`, you cannot call `inputDirectory.set(copyTask.flatMap { it.destinationDir })` right now.

**Comment by bowserf** [comment]
Hello,

I had the same problem with no way to lazy set a destinationDir, and I didn't want to wait for the Gradle 9 release.
I looked at the Copy task code and didn't find this limitation necessary. We could easily rewrite it to use a DirectoryProperty as the output.

Perhaps my code is completely wrong, but it seems to work for my needs. Hopefully, it might help others!
If someone could provide feedback in case I've missed something significant, it would be greatly appreciated =)

Link to the [gist](https://gist.github.com/bowserf/d5b69eec1f130af31eed44b3c2b8c687)

```kotlin
import org.gradle.api.file.Directory
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.internal.file.copy.CopyAction
import org.gradle.api.internal.file.copy.CopySpecInternal
import org.gradle.api.internal.file.copy.DestinationRootCopySpec
import org.gradle.api.internal.file.copy.FileCopyAction
import org.gradle.api.provider.Provider
import org.gradle.api.tasks.AbstractCopyTask
import org.gradle.api.tasks.OutputDirectory
import org.gradle.work.DisableCachingByDefault

/**
 * Kotlin implementation of the "Copy" Gradle task which has a DirectoryProperty as output (the
 * property allows to set the directory later) instead of having to know the "destination dir"
 * during configuration step, like in the "Copy" task.
 *
 * The "Copy" task code can be found here:
 * https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/tasks/Copy.java
 */
@DisableCachingByDefault(because = "Not worth caching")
abstract class LazyCopy : AbstractCopyTask() {

    private var lazyDestinationDirectory: DirectoryProperty

    init {
        val objectFactory = project.objects
        lazyDestinationDirectory = objectFactory.directoryProperty()
    }

    fun setDestinationDirectory(destinationDir: Provider<Directory>) {
        lazyDestinationDirectory.set(destinationDir)
    }

    @OutputDirectory
    fun getDestinationDirectory(): DirectoryProperty {
        return lazyDestinationDirectory
    }

    override fun createCopyAction(): CopyAction {
        val fileResolver = fileLookup.getFileResolver(getDestinationDirectory().get().asFile)
        return FileCopyAction(fileResolver)
    }

    override fun createRootSpec(): CopySpecInternal {
        return project.objects.newInstance(
            DestinationRootCopySpec::class.java, super.createRootSpec()
        )
    }

    override fun getRootSpec(): DestinationRootCopySpec {
        return super.getRootSpec() as DestinationRootCopySpec
    }
}
```

**Comment by asodja** [comment]
We plan to migrate `Copy#destinationDir` to DirectoryProperty in Gradle 9.

Custom implementation is ok. I would just recommend to not extend `AbstractCopyTask`, but rather extend `DefaultTask` and inject [FileSystemOperations](https://github.com/gradle/gradle/blob/0ec19312f166cf079c38131d4744915f5863f5b9/subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemOperations.java#L35) in to the task. Such pattern favors composition over inheritence. We see that as a better approach since task author has more control over task behaviour and it's inputs/outputs, and such code is less susceptible to API changes.

**Comment by bowserf** [comment]
Thank you so much for your quick feedback!

I tried to find a solution to the problem as fast as possible, that's why I mirrored the code of the current Copy task with the minimum number of changes, but I will definitively take the time to apply your valuable comments 🙏

**Comment by big-guy** [comment]
This has been deferred to 10.0.

As a couple of workarounds:
To lazily assign the destination, you can use `into`:
```
copy.into(provider)
```
To get the destination lazily, you can do a few things:
1. Use `ConfigurableFileCollection.from`, this doesn't require calling `copy.destinationDir` at all.
2. Use `DirectoryProperty.fileProvider`, `consumer.fileProvider(copy.map { it.destinationDir })` as long as copy is a `TaskProvider<Copy>`.

**Comment by Goooler** [comment]
Duplicate of #13275
